### PR TITLE
fix add and drop fk relationships with involved primary key columns not named 'id'

### LINKED ISSUE #20536: drop relationship column not working for id columns not named "id"
##### **Overview of the issue**
When generating a liquibase changelog dropping a foreign key relationship, the changelog generated does not consider id columns not named 'id':

https://github.com/jhipster/generator-jhipster/blob/4e52f15041beee04ffce42ff3be3f542661ada8b/generators/liquibase-changelogs/templates/src/main/resources/config/liquibase/changelog/updated_entity.xml.ejs#L148

When applying such a changelog, that fails, leading to a not starting application:
2022-12-07T09:33:49.571+01:00 ERROR 23536 --- [  restartedMain] o.s.boot.SpringApplication               : Application run failed 

```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'liquibase' defined in class path resource [de/mt/dataview/config/LiquibaseConfiguration.class]: Invocation of init method failed; nested exception is liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset config/liquibase/changelog/20221207073604_updated_entity_StreamError.xml::20221207073604-1-drop-relationships::jhipster:
     Reason: liquibase.exception.DatabaseException: Can't DROP 'stream_data_id'; check that column/key exists [Failed SQL: (1091) ALTER TABLE fx.stream_error DROP COLUMN stream_data_id]
```

##### **Motivation for or Use Case**
Applying changelog fails and changelog must be modified manually.

##### **Suggest a Fix**
https://github.com/jhipster/generator-jhipster/blob/4e52f15041beee04ffce42ff3be3f542661ada8b/generators/liquibase-changelogs/templates/src/main/resources/config/liquibase/changelog/updated_entity.xml.ejs#L148 must be adjusted

##### **JHipster Version(s)**
7.9.3

##### **Browsers and Operating System**

- [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by OmarHawk**
Fix #20536

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by OmarHawk** [comment]
I'll continue working on this. ;-)

**Comment by mshima** [comment]
- Maybe a snapshot test for relationship removal (just a snapshot is enough IMO):
https://github.com/jhipster/generator-jhipster/blob/5f645ea28f33ec1aa2e249774e82ff3b29e66ef4/generators/liquibase-changelogs/incremental-liquibase.spec.mts#L592

- And an actual relationship removal at https://github.com/jhipster/generator-jhipster/tree/main/test-integration/incremental-changelog-samples.

**Review by mshima — DISMISSED** [review]

**Comment by OmarHawk** [comment]
> * Maybe a snapshot test for relationship removal (just a snapshot is enough IMO):
>   https://github.com/jhipster/generator-jhipster/blob/5f645ea28f33ec1aa2e249774e82ff3b29e66ef4/generators/liquibase-changelogs/incremental-liquibase.spec.mts#L592

I'll add one

> * And an actual relationship removal at https://github.com/jhipster/generator-jhipster/tree/main/test-integration/incremental-changelog-samples.

I added `@Id` fields where there is already a relationship removal in the existing test case. It is named 'rename field' but relationship removal is also already in there... Maybe, we should rename that test. ;-)

**Comment by OmarHawk** [comment]
> > * Maybe a snapshot test for relationship removal (just a snapshot is enough IMO):
> >   https://github.com/jhipster/generator-jhipster/blob/5f645ea28f33ec1aa2e249774e82ff3b29e66ef4/generators/liquibase-changelogs/incremental-liquibase.spec.mts#L592
> 
> I'll add one

Done.

File: generators/liquibase-changelogs/templates/src/main/resources/config/liquibase/changelog/updated_entity.xml.ejs [hunk 1]
```diff
@@ -78,9 +78,12 @@
         <addColumn tableName="<%= entity.entityTableName %>">
             <%_
             for (relationship of addedRelationships) {
-                relationshipColumnType = relationship.otherEntityName === 'user' && authenticationTypeOauth2 ? 'varchar(100)' : 'bigint';
                 if (relationship.shouldWriteRelationship) { _%>
-                <column name="<%= relationship.columnName %>_id" type="<%= relationshipColumnType %>"/>
+                    <%_ for (field of relationship.otherEntity.primaryKey.fields) {
```
Comment (mshima) [inline generators/liquibase-changelogs/templates/src/main/resources/config/liquibase/changelog/updated_entity.xml.ejs:82]:
```suggestion
                    <%_ for (const otherEntityPkField of relationship.otherEntity.primaryKey.fields) { 
```

File: generators/liquibase-changelogs/templates/src/main/resources/config/liquibase/changelog/updated_entity.xml.ejs [hunk 2]
```diff
@@ -78,9 +78,12 @@
         <addColumn tableName="<%= entity.entityTableName %>">
             <%_
             for (relationship of addedRelationships) {
-                relationshipColumnType = relationship.otherEntityName === 'user' && authenticationTypeOauth2 ? 'varchar(100)' : 'bigint';
                 if (relationship.shouldWriteRelationship) { _%>
-                <column name="<%= relationship.columnName %>_id" type="<%= relationshipColumnType %>"/>
+                    <%_ for (field of relationship.otherEntity.primaryKey.fields) { 
+                        let relationshipColumnType = relationship.otherEntityName === 'user' && authenticationTypeOauth2 ? 'varchar(100)' : field.columnType;
+                    _%>
+                <column name="<%= relationship.columnName %>_<%= field.columnName %>" type="<%= relationshipColumnType %>"/>
```
Comment (mshima) [inline generators/liquibase-changelogs/templates/src/main/resources/config/liquibase/changelog/updated_entity.xml.ejs:85]:
```suggestion
                <column name="<%= relationship.columnName %>_<%= otherEntityPkField.columnName %>" type="<%= otherEntityPkField.columnType %>"/>
```

File: generators/liquibase-changelogs/templates/src/main/resources/config/liquibase/changelog/updated_entity.xml.ejs [hunk 3]
```diff
@@ -78,9 +78,12 @@
         <addColumn tableName="<%= entity.entityTableName %>">
             <%_
             for (relationship of addedRelationships) {
-                relationshipColumnType = relationship.otherEntityName === 'user' && authenticationTypeOauth2 ? 'varchar(100)' : 'bigint';
                 if (relationship.shouldWriteRelationship) { _%>
-                <column name="<%= relationship.columnName %>_id" type="<%= relationshipColumnType %>"/>
+                    <%_ for (field of relationship.otherEntity.primaryKey.fields) { 
+                        let relationshipColumnType = relationship.otherEntityName === 'user' && authenticationTypeOauth2 ? 'varchar(100)' : field.columnType;
```
Comment (mshima) [inline generators/liquibase-changelogs/templates/src/main/resources/config/liquibase/changelog/updated_entity.xml.ejs:83]: User should have the primary key fields data too.
```suggestion
```

**Review by mshima — COMMENTED** [review]
@OmarHawk can you test this change?

**Review by mshima — APPROVED** [review]

**Comment by mshima** [comment]
Thanks @OmarHawk

**Comment by OmarHawk** [comment]
Bug bounty claim https://opencollective.com/generator-jhipster/expenses/120027 :-)

**Comment by DanielFran** [comment]
@OmarHawk Approved

**Pull request merged**